go/vt/vttablet/tabletmanager/vreplication/vcopier_copy_all.go

```diff
@@ -0,0 +1,288 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package vreplication
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"strconv"
+	"time"
+
+	"vitess.io/vitess/go/vt/sqlparser"
+
+	"google.golang.org/protobuf/encoding/prototext"
+
+	"vitess.io/vitess/go/sqltypes"
+	"vitess.io/vitess/go/vt/binlog/binlogplayer"
+	"vitess.io/vitess/go/vt/log"
+	binlogdatapb "vitess.io/vitess/go/vt/proto/binlogdata"
+	querypb "vitess.io/vitess/go/vt/proto/query"
+	"vitess.io/vitess/go/vt/vterrors"
+)
+
+/*
+This file is similar to vcopier.go: it handles the copy phase for the ForeignKeyMode where all tables
+are streamed in a single phase.
+*/
+
+type copyAllState struct {
+	vc               *vcopier
+	plan             *ReplicatorPlan
+	currentTableName string
+	tables           map[string]bool
+}
+
+func newCopyAllState(vc *vcopier) (*copyAllState, error) {
+	state := &copyAllState{
+		vc: vc,
+	}
+	plan, err := buildReplicatorPlan(vc.vr.source, vc.vr.colInfoMap, nil, vc.vr.stats)
+	if err != nil {
+		return nil, err
+	}
+	state.plan = plan
+	state.tables = make(map[string]bool, len(plan.TargetTables))
+	for _, table := range plan.TargetTables {
+		state.tables[table.TargetName] = false
+	}
+	return state, nil
+}
+
+func (vc *vcopier) copyAll(ctx context.Context, settings binlogplayer.VRSettings) error {
+	var err error
+
+	log.Infof("Starting copyAll for %s", settings.WorkflowName)
+	defer log.Infof("Returning from copyAll for %s", settings.WorkflowName)
+	defer vc.vr.dbClient.Rollback()
+
+	state, err := newCopyAllState(vc)
+	if err != nil {
+		return err
+	}
+
+	ctx, cancel := context.WithTimeout(ctx, copyPhaseDuration)
+	defer cancel()
+
+	rowsCopiedTicker := time.NewTicker(rowsCopiedUpdateInterval)
+	defer rowsCopiedTicker.Stop()
+	copyStateGCTicker := time.NewTicker(copyStateGCInterval)
+	defer copyStateGCTicker.Stop()
+
+	parallelism := getInsertParallelism()
+	copyWorkerFactory := vc.newCopyWorkerFactory(parallelism)
+	var copyWorkQueue *vcopierCopyWorkQueue
+
+	// Allocate a result channel to collect results from tasks.
+	resultCh := make(chan *vcopierCopyTaskResult, parallelism*4)
+	defer close(resultCh)
+
+	var lastpk *querypb.Row
+	var pkfields []*querypb.Field
+	var lastpkbv map[string]*querypb.BindVariable
+	// Use this for task sequencing.
+	var prevCh <-chan *vcopierCopyTaskResult
+	var gtid string
+
+	serr := vc.vr.sourceVStreamer.VStreamTables(ctx, func(resp *binlogdatapb.VStreamTablesResponse) error {
+		defer vc.vr.stats.PhaseTimings.Record("copy", time.Now())
+		defer vc.vr.stats.CopyLoopCount.Add(1)
+		log.Infof("VStreamTablesResponse: received table %s, #fields %d, #rows %d, gtid %s, lastpk %+v",
+			resp.TableName, len(resp.Fields), len(resp.Rows), resp.Gtid, resp.Lastpk)
+		tableName := resp.TableName
+		gtid = resp.Gtid
+
+		update := binlogplayer.GenerateUpdateRowsCopied(vc.vr.id, vc.vr.stats.CopyRowCount.Get())
+		_, _ = vc.vr.dbClient.Execute(update)
```

Comment:
Do we really want to ignore any errors here? I wouldn't think so. I get that it's only updating metadata though... and I guess we don't care too much in this case. So maybe it's fine. If nothing else it's worth a comment. :-)